plugin/db/clickhouse/clickhouse.go

```diff
@@ -407,14 +407,14 @@ func (driver *Driver) SetupMigrationIfNeeded(ctx context.Context) error {
 	return nil
 }
 
-// CheckOutOfOrderVersion will return versions that are higher than the given version.
-func (Driver) CheckOutOfOrderVersion(ctx context.Context, tx *sql.Tx, namespace string, source db.MigrationSource, version string) (minVersionIfValid *string, err error) {
+// CheckOutOfOrderVersion will return the version that is higher than the given version since the last baseline.
```

Comment:
I feel the overall logic becomes quite complex. Let's take a step back and review the problem. I think there are 2 problems we want to solve here.

1)  Make sure the new version is larger than the largest version since the last baseline or if no baseline exists, since the very beginning.

2)  If the existing migration record history is already out of order, how to live with that anomaly and apply the migration as long as it meets 1.

So to achieve both, the new migration record should have both the largest sequence and the largest version number since the last baseline or if no baseline exists, since the very beginning.

Then the problem comes to:

a. We fetch the existing largest sequence since the last baseline or if no baseline exists, since the very beginning. and use (largest sequence + 1) as the new sequence

b. We fetch the existing largest version since the last baseline or if no baseline exists, since the very beginning.

So the actual CheckOutOfOrderVersion logic is a simple string comparison. The real complexity involving SQL is in the above 2 steps.

for a), the query would be something like

SELECT max(id) FROM xxx WHERE type = 'BASELINE' OR 'BRANCH'

if max id exists then start the search since that ID to find the max sequence. If max id does not exist, then we find the max sequence of all migration records.

b) is very similiar.

Reply:
Yeah, I think this logic is doing what above intends to do. minSequence is the sequence number of previous baseline/branch.

Reply:
The changed logic is 

1. We get the largest sequence number

2. We get the corresponding version string for that largest sequence number. And compare that version with the new specified version

The flaw is in step 2, because the version for that largest sequence number is NOT necessarily the largest version.

Reply:
The current logic should do it.